### PR TITLE
Solace Read connector: adding implementations of SempClient and SempClientFactory

### PR DESCRIPTION
**Pull request by bzablocki**
This is the 4th PR for the Solace Read connector (addresses #31440).
It depends on the #31541, #31539 and #31476 (root).

The files that are added/modified in this PR, comparing to the #31541:
- build.gradle
- BasicAuthSempClient.java
- BasicAuthSempClientFactory.java
- BrokerResponse.java
- SempBasicAuthClientExecutor.java
- Semp.java
- package-info.java
- SerializableSupplier.java
- SempBasicAuthClientExecutorTest.java

These are implementations  of SempClient and SempClientFactory. SempClientFactory can be used as an argument to `SolaceIO.Read.withSempClientFactory(...)`. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by bzablocki** [comment]
assign set of reviewers

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthSempClientFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    public abstract BasicAuthSempClientFactory build();
+  }
+
+  @Override
+  public SempClient create() {
+    return new BasicAuthSempClient(
+        host(), username(), password(), vpnName(), getHttpRequestFactorySupplier());
+  }
+
+  @SuppressWarnings("return")
```
Comment (Abacn) [inline sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthSempClientFactory.java:77]: a brief comment for the justification of suppress checkFramework violation, or fix it

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+  private HttpResponse executePost(GenericUrl url, ImmutableMap<String, Object> parameters)
+      throws IOException {
+    HttpContent content = new JsonHttpContent(GsonFactory.getDefaultInstance(), parameters);
+    HttpRequest request = requestFactory.buildPostRequest(url, content);
+    return execute(request);
+  }
+
+  private HttpResponse execute(HttpRequest request) throws IOException {
+    request.setNumberOfRetries(2);
```
Comment (Abacn) [inline sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:128]: what consideration here for retry twice? Good to have comment for occurrence of hardcoded numbers

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    HttpResponse response;
+    try {
+      response = request.execute();
+    } catch (HttpResponseException e) {
+      if (authFromCookie && e.getStatusCode() == 401) {
+        COOKIE_MANAGER.getCookieStore().removeAll();
+        // execute again without cookies to refresh the token.
+        return execute(request);
+      } else {
```
Comment (Abacn) [inline sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:147]: there may be other retriable return codes? e.g. from the spec does it possible to return server side error 50X, rate limiting 429, etc. Fine to leave as is for now but good to comment here if we do want to handle them in the future what should be done

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    } else {
+      httpHeaders.setBasicAuthentication(username, password);
+      request.setHeaders(httpHeaders);
+    }
+
+    HttpResponse response;
+    try {
+      response = request.execute();
+    } catch (HttpResponseException e) {
+      if (authFromCookie && e.getStatusCode() == 401) {
```
Comment (Abacn) [inline sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:143]: is this (make a request and check for 401) the only way to refresh token?

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+      throws IOException {
+    HttpContent content = new JsonHttpContent(GsonFactory.getDefaultInstance(), parameters);
+    HttpRequest request = requestFactory.buildPostRequest(url, content);
+    return execute(request);
+  }
+
+  private HttpResponse execute(HttpRequest request) throws IOException {
+    request.setNumberOfRetries(2);
+    HttpHeaders httpHeaders = new HttpHeaders();
+    boolean authFromCookie = COOKIE_MANAGER.getCookieStore().getCookies().size() > 0;
```
Comment (Abacn) [inline sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:130]: The COOKIE_MANAGER is worker wide, suppose there are two SolaceIO sources from different servers, could there be conflict? Should the static COOKIE_MANAGER  be a map keyed with destination?

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/utils/SerializableSupplier.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.solace.utils;
+
+import java.io.Serializable;
+
+@FunctionalInterface
+public interface SerializableSupplier<OutputT> extends Serializable {
```
Comment (Abacn) [inline sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/utils/SerializableSupplier.java:23]: Is this the same as https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dofn/SerializableSupplier.java ?

Probably move the definition of this interface to java core util package (https://github.com/apache/beam/tree/master/sdks/java/core/src/main/java/org/apache/beam/sdk/util) and use that one?

**Review by Abacn — COMMENTED** [review]
Thanks, left a few comments

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/utils/SerializableSupplier.java (hunk 6 above)
Comment (bzablocki) [reply sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/utils/SerializableSupplier.java:23]: Makes sense, I created #31766

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthSempClientFactory.java (hunk 1 above)
Comment (bzablocki) [reply sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/BasicAuthSempClientFactory.java:77]: Thanks, I can actually remove it from here.

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java (hunk 2 above)
Comment (bzablocki) [reply sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:128]: I extracted the variable and added a comment.

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java (hunk 3 above)
Comment (bzablocki) [reply sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:147]: Good point, I'll leave a comment to consider handling it in the future

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java (hunk 4 above)
Comment (bzablocki) [reply sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:143]: The Solace [documentation](https://docs.solace.com/Admin/SEMP/SEMP-Security.htm) lacks guidance on detecting token expiration. Typically, tokens include an `expires_in` field for calculating refresh times. However, Solace seems to omit this information. Currently, the only way to check token validity is by sending a request and looking for a 401 Unauthorized response, which would then trigger token refresh.

File: sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java (hunk 5 above)
Comment (bzablocki) [reply sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/broker/SempBasicAuthClientExecutor.java:130]: Good point, changed to a ConcurrentHashMap to make sure it's thread-safe.

**Review by Abacn — APPROVED** [review]

**Pull request merged**